src/python/pants/backend/adhoc/code_quality_tool.py

```diff
@@ -0,0 +1,549 @@
+    alias: ClassVar[str] = "args"
+    default = ()
+    help = help_text(
+        lambda: f"""
+        Extra arguments to pass into the `{CodeQualityToolRunnableField.alias}` field
+        before the list of source files
+        """
+    )
+
+
+class CodeQualityToolExecutionDependenciesField(SpecialCasedDependencies):
+    alias: ClassVar[str] = "execution_dependencies"
+    required = False
+    default = None
+
+    help = help_text(
+        """
+        Additional dependencies that need to be available when running the tool.
+        Typically used to point to config files.
+        """
+    )
+
+
+class CodeQualityToolRunnableDependenciesField(SpecialCasedDependencies):
+    alias: ClassVar[str] = "runnable_dependencies"
+    required = False
+    default = None
+
+    help = help_text(
+        lambda: f"""
+        The runnable dependencies for this command.
+
+        Dependencies specified here are those required to exist on the `PATH` to make the command
+        complete successfully (interpreters specified in a `#!` command, etc). Note that these
+        dependencies will be made available on the `PATH` with the name of the target.
+
+        See also `{CodeQualityToolExecutionDependenciesField.alias}`.
+        """
+    )
+
+
+class CodeQualityToolTarget(Target):
+    alias: ClassVar[str] = "code_quality_tool"
+    core_fields = (
+        *COMMON_TARGET_FIELDS,
+        CodeQualityToolRunnableField,
+        CodeQualityToolArgumentsField,
+        CodeQualityToolExecutionDependenciesField,
+        CodeQualityToolRunnableDependenciesField,
+        CodeQualityToolFileGlobIncludeField,
+        CodeQualityToolFileGlobExcludeField,
+    )
+
+    help = help_text(
+        lambda: f"""
+        Configure a runnable to use as a linter, fixer or formatter
+
+        Example BUILD file:
+
+            {CodeQualityToolTarget.alias}(
+                {CodeQualityToolRunnableField.alias}=":flake8_req",
+                {CodeQualityToolExecutionDependenciesField.alias}=[":config_file"],
+                {CodeQualityToolFileGlobIncludeField.alias}=["**/*.py"],
+            )
+        """
+    )
+
+
+@dataclass(frozen=True)
+class CodeQualityToolAddressString:
+    address: str
+
+
+@dataclass(frozen=True)
+class CodeQualityTool:
+    runnable_address_str: str
+    args: tuple[str, ...]
+    execution_dependencies: tuple[str, ...]
+    runnable_dependencies: tuple[str, ...]
+    file_glob_include: tuple[str, ...]
+    file_glob_exclude: tuple[str, ...]
+    target: Target
+
+
+@rule
+async def find_code_quality_tool(request: CodeQualityToolAddressString) -> CodeQualityTool:
+    tool_address = await Get(
+        Address,
+        AddressInput,
+        AddressInput.parse(request.address, description_of_origin="code quality tool target"),
+    )
+
+    addresses = Addresses((tool_address,))
+    addresses.expect_single()
+
+    linter_targets = await Get(Targets, Addresses, addresses)
+    target = linter_targets[0]
```

Comment:
I am not sure. I think that since we fetched an address and already asserted it to be a single one that there would necessarily be a target. In `adhoc_tool` we do a similar unchecked `[0]` of a target after previously asserting the existence of the address

https://github.com/pantsbuild/pants/blob/e7ed2092b2cc64673c5aba5b6400f48a3775ea2e/src/python/pants/backend/adhoc/adhoc_tool.py#L74-L91